File: source/Chronos/Chronos.Agent/UnitBase.cpp
#include "stdafx.h"
#include "Chronos.Agent.h"

namespace Chronos
{
	namespace Agent
	{
		UnitBase::UnitBase()
			: _beginLifetime(0), _endLifetime(0)
		{
			Uid = 0;
			Id = 0;
			_name = null;
		}

		UnitBase::~UnitBase()
		{
			__FREEOBJ(_name);
		}

		void UnitBase::Initialize(__uint uid, __uptr id, __uint beginLifetime)
		{
			Uid = uid;
			Id = id;
			_beginLifetime = beginLifetime;
		}

		void UnitBase::Close(__uint endLifetime)
		{
			_endLifetime = endLifetime;
		}
		
		__uint UnitBase::GetBeginLifetime()
		{
			return _beginLifetime;
		}

		__uint UnitBase::GetEndLifetime()
		{
			return _endLifetime;
		}

		__string* UnitBase::GetName()
		{
			return _name;
		}

		__bool UnitBase::GetIsAlive()
		{
			return _endLifetime <= 0;
		}
	}
}

File: source/Chronos.Common/Chronos.Common/_Units/UnitBase.cs
using System;

namespace Chronos.Common
{
    [Serializable]
    public abstract class UnitBase
    {
        private NativeUnitBase _nativeUnit;

        protected UnitBase(NativeUnitBase nativeUnit)
        {
            _nativeUnit = nativeUnit;
        }

        protected internal NativeUnitBase NativeUnit
        {
            get { return _nativeUnit; }
        }

        public ulong Uid
        {
            get { return NativeUnit.Uid; }
        }

        public ulong Id
        {
            get { return NativeUnit.Id; }
        }

        public ulong BeginLifetime
        {
            get { return NativeUnit.BeginLifetime; }
        }

        public ulong EndLifetime
        {
            get { return NativeUnit.EndLifetime; }
        }

        public string Name
        {
            get { return NativeUnit.Name; }
        }

        protected internal void Update(NativeUnitBase unitBase)
        {
            _nativeUnit = unitBase;
        }
    }
}
